### ISSUE TITLE
Zero-initializers for immutable types?

### ISSUE THREAD

**Issue by palmskog** [Microsoft/dafny#191]
In the new game with non-null types and zero-initializer types, it seems impossible to use standard container classes for storing most immutable types such as custom algebraic datatypes, `seq<int>`, and `set<bool>` (but not `int`, `bool`, or `real`).

As one example, the type parameter in an [implementation](https://gist.github.com/palmskog/9767ca96d19d8887c5de28fe5404c21e) of an array-backed list (ArrayList as in Java collections) needs to be `T(0)`, but this prevents having, e.g., `ArrayList<string>`.

As another example, an implementation of the `TryGetValue` [method](https://msdn.microsoft.com/en-us/library/bb299639(v=vs.110).aspx) in a dictionary/map class seemingly requires the value type to be `V(0)` since a value must always be returned. Again, this makes the container impossible to use for many interesting types. (In addition, I haven't found any way to explicitly reference the zero-initializer for a type.)

Am I missing something, or are there any plans for making immutable types besides `int`, `bool`, and `real` zero-initializable? And is there any recommended workaround besides axiomatization for storing immutable types in containers?

**Comment by RustanLeino** [comment]
I wrestled a lot with this language design decision, and it seems the struggle is not over. The "benchmark example" I had in mind when wrestling with the issue was indeed the kind of `Arraylist` you're bringing up. I'll give my change suggestions below. But first, a few remarks about zero initialization.

## Issues with initialization

Dafny does know how to initialize variables of type `string`. Type `string` is a synonym for `seq<char>` and Dafny knows how to initialize any `seq`, and indeed any collection type (`set`, `multiset`, ...). It also knows how to initialize values of many inductive and co-inductive datatypes. For example, you can have an uninitialized out-parameter of one of these types and Dafny is happy.

```
method VickenSomHelst() returns (s: string)
{
  // Look, Ma! No manual initialization of s.
}
```

Unfortunately, this does not carry over to type parameters. More precisely, there's a difference between knowing how to initialize variables of a known type (like `string` in my example above) and knowing how to initialize variables whose type is a type parameter (like in your `ArrayList<T>`). So, at the point where you'd attempt a `new ArrayList<string>()`, it is known how to initialize a `string`, but on the other side of this call, where the type is known just as `T`, it is not known how to initialize a `T`. :(

Using some run-time information about the instantiations of type parameters would seem to help. After all, C# and the .NET virtual machine support an expression `default(T)`, which also works when `T` is a type parameter. The problem is that the mapping of Dafny types to .NET types is not one-to-one. For example, `int` and `nat` are compiled to the same type, and so are the following two subset types:

```
type EvenInt = x | x % 2 == 0
type OddInt = x | x % 2 == 1 witness 1
```

To obtain the identity of these types at run time would require the Dafny compiler to generate different .NET types for these, which leads to other issues. Since the mapping of Dafny types to .NET types is not one-to-one is also why Dafny can't provide you with an expression for getting an arbitrary value of a any type.

I've considered several ways around this problem, but each has left me unhappy in some way. As I'm writing this, I'm thinking of a new possible compromise, which is to use reflection at run time to support initialization for _some_ of the types that the compiler would know how to initialize if it only knew what the type was, with the restriction of doing this only for one of the Dafny types that map to the same underlying type. Let me give this some thought. Other suggestions are welcome, too.

## Suggestions for ArrayList

Here is a way you can change `ArrayList<T(0)>` back to `ArrayList<T>`. As I see it, the biggest issue is the `InitialCapacity` constructor.

In the anonymous constructor, don't allow an array of size 10, but instead allocate an array of size 0.

Add another parameter to the `InitialCapacity` constructor, forcing the caller to pass in an example element of the type `T`.

```
    constructor InitialCapacity(initialCapacity: int, archetype: T)
      ...
    {
      ...
      arr := new T[initialCapacity](_ => archetype);
      ...
    }
```

Do the same with `EnsureCapacity`. In this case, you can use the built-in array initialization to also do what your call to `ArrayCopy` was doing.

```
    method EnsureCapacity(minCapacity: int, archetype: T)
    ...
    {
      ...
        // BEFORE:  var newArr := new T[newCapacity];
        // BEFORE:  ArrayCopy(arr, 0, newArr, 0, size);
        // NOW:
        var size', arr' := size, arr;
        var newArr := new T[newCapacity](i reads arr' requires 0 <= i => if i < size' then arr'[i] else archetype);
      ...
    }
```

You'll have to fix up the calls to `EnsureCapacity`, but I imagine that's limited to the two calls inside the class. Here, the caller `Add` conveniently already has an element to supply to `EnsureCapacity`:

```
    method Add(element: T) returns (index: int)
      ...
    {
      EnsureCapacity(size + 1, element);
      ...
    }
```

The other call is `Append`. By changing the code to do nothing if the `ArrayList` parameter has no elements, you can easily fix this call, too:

```
    method Append(a: ArrayList<T>)
      ...
    {
      if a.size != 0 {
        ...
        EnsureCapacity(size + numNew, a.arr[0]);
        ...
      }
    }
```

For your envisioned `TryGetValue` method, you can declare an option type:

```
datatype Option<T> = Some(get: T) | None
```

and use it as the return value. (That's probably what the Java designers would have done, too, if their language had supported option types.)

```
method TryGetValue(index: int) returns (element: Option<T>)
  requires Valid()
  ensures element == if 0 <= index < size then Some(list[index]) else None
{
  element := if 0 <= index < size then Some(arr[index]) else None;
}
```

## Miscellaneous comment about constructors

For a long time, Dafny's constructors used to be like other methods in that they were called after the receiver object had been initialized. This meant you had to have a `modifies this` clause on every constructor, which is unnatural. It also meant you had to exclude `this` from the fresh things that the postcondition promises `Repr` to contain. In a recent Dafny version, this finally changed and the constructor is now, conceptually, what allocates the new object. Consequently, `this` is fresh, too, which means that instead of the postcondition `fresh(Repr - {this})`, you can simply write `fresh(Repr)`.

Program safely,
  Rustan

**Comment by palmskog** [comment]
@RustanLeino Thanks so much for the detailed explanation and workarounds. For anyone else reading this, I applied Rustan's advice and obtained an [updated ArrayList](https://gist.github.com/palmskog/04673cef891e09732b0540c8578aa5de) the verifies without requiring `(0)` in the type parameter.

I'll leave up to the maintainers whether this issue should be closed - it might be good to have it open as a reference for Dafny users until the reference manual covers it, or a more general solution (e.g., reflection) is merged.

**Comment by RustanLeino** [comment]
One more general suggestion. If you want, you can implement the `Remove` method with Dafny's `forall` statement instead of using a loop. This means you don't need to supply a loop invariant (because the `forall` statement is an aggregate statement, not a loop).

```
method Remove(index: int)
  ...
{
  forall i | index + 1 <= i < size {
    arr[i - 1] := arr[i];
  }
  size := size - 1;
  list := arr[..size];
}
```

**Comment by palmskog** [comment]
Tack igen! I'm aware of the `forall` statement, but the `.cs` code that gets spilled from it is nearly incomprehensible to me. Are the non-functional characteristics (performance, memory use) usually the same for loops vs. `forall` when compiled?

Also, for the benefit of other readers, I should link to a map [implementation](https://gist.github.com/palmskog/f957cf4c587cb753d138f945f214e4e2) with Rustan's suggestion about the option type incorporated.

**Comment by RustanLeino** [comment]
I'm glad to hear my comments were helpful. Meanwhile, I implemented the idea I had to admit more types as having a zero initializer (as far as Dafny is concerned). Thanks for prompting me to do so.

I appreciate what you said about leaving this Issue open, so that others may learn from the discussion. However, with this new feature (and the suggestion of trying to implement the class without the `(0)`-restriction), I feel the Issue itself has been addressed and can be closed. To let others learn from the discussion, feel free to post your original question (or some variation thereof) on StackOverflow and giving it the "dafny" tag. I can then copy my answer (or some variation thereof) there.

  Rustan
